mautic/mautic
# Company Import

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | Pending
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR adds the ability to import companies in the same way that we do contacts. You can upload your company data via CSV, map the fields to their mautic equivalent, then import. You can do this with just company data or along with your contact info. 

When you upload a contact CSV, you now have the ability to map any of your data to Company fields within Mautic. If an associated company is found, then that company will be updated and the contact will be associated with that company. If no company is found then it will be created (provided that there is enough information) and the contact assigned.

Finally, this updates the contact/company import by flipping the way importing is mapped. Previously, all Mautic fields would be listed out on the import page with your CSV fields in the dropdowns to select from. This made for very lengthy import mapping pages. That assignment is now flipped - The fields displayed on the import page are the column headers found in your CSV, and the dropdowns are the mautic fields for you to map to. We feel this will give a much better user experience when importing a select number of fields.

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Apply PR
2. Create a CSV with company data
3. Go to Companies page and see the new dropdown button next to `New`. Select `Import`. 
4. Upload your CSV, then map the headers to the appropriate Mautic field.
5. Run the import
6. See your imported company data
7. Create another CSV, this time containing contact AND company data. 
8. Run the import, and ensure the companies are created/updated and that contacts are assigned to their appropriate company.
9. Do all of this with CLI imports as well.
10. Ensure that only company imports are listed on the company import page and that only contact imports are displayed on the contact import page.

#### List deprecations along with the new alternative:
1. `LeadModel::importLead` has been deprecated and replaced with `LeadModel::import`. It has the same method signature. `importLead` will be removed in 3.0

#### List backwards compatibility breaks:
1. Imports that are created/mapped prior to this PR being applied will now fail to import. This is due to the field mapping change. You will need to reupload/remap those imports if the import has not run prior to applying this PR.
